modules/siddhi-core/src/main/java/org/wso2/siddhi/core/aggregation/AggregationRuntime.java

```diff
@@ -93,18 +115,68 @@ public EntryValveExecutor getEntryValveExecutor() {
     }
 
     public StreamEvent find(StateEvent matchingEvent, CompiledCondition compiledCondition) {
-        return null;
-        //todo implement
+        String perValueAsString = perExpressionExecutor.execute(matchingEvent).toString();
+        TimePeriod.Duration perValue = TimePeriod.Duration.valueOf(perValueAsString.toUpperCase());
+        if (!incrementalExecutorMap.keySet().contains(perValue)) {
+            throw new SiddhiAppRuntimeException("The aggregate values for " + perValue.toString()
+                    + " granularity cannot be provided since aggregation definition " + aggregationDefinition.getId()
+                    + " does not contain " + perValue.toString() + " duration");
+        }
+        return ((IncrementalAggregateCompileCondition) compiledCondition).find(matchingEvent, perValue,
+                incrementalExecutorMap.get(perValue).isRoot());
     }
 
+    public IncrementalAggregateCompileCondition compileExpression(Expression expression, Within within, Expression per,
+            MatchingMetaInfoHolder matchingMetaInfoHolder, List<VariableExpressionExecutor> variableExpressionExecutors,
+            Map<String, Table> tableMap, String queryName, SiddhiAppContext siddhiAppContext) {
+        Map<TimePeriod.Duration, CompiledCondition> tableCompiledConditions = new HashMap<>();
+        perExpressionExecutor = ExpressionParser.parseExpression(per, matchingMetaInfoHolder.getMetaStateEvent(),
+                matchingMetaInfoHolder.getCurrentState(), tableMap, variableExpressionExecutors, siddhiAppContext,
+                false, 0, queryName);
+        if (perExpressionExecutor.getReturnType() != Attribute.Type.STRING) {
+            throw new SiddhiAppCreationException("Query " + queryName + "'s per value expected a string but found "
+                    + perExpressionExecutor.getReturnType());
+        }
+        Expression withinExpression = Expression.incrementalWithinTime(within, Expression.variable("_TIMESTAMP"));
+        for (Map.Entry<TimePeriod.Duration, Table> entry : aggregationTables.entrySet()) {
+            CompiledCondition tableCompileCondition = entry.getValue().compileCondition(withinExpression,
+                    newMatchingMetaInfoHolder(matchingMetaInfoHolder, entry.getValue().getTableDefinition()),
+                    siddhiAppContext, variableExpressionExecutors, tableMap, queryName);
+            tableCompiledConditions.put(entry.getKey(), tableCompileCondition);
+
+            inMemoryStoreMap.put(entry.getKey(), incrementalExecutorMap.get(entry.getKey()).getInMemoryStore());
+        }
+        CompiledCondition inMemoryStoreCompileCondition = OperatorParser.constructOperator(
+                new ComplexEventChunk<>(true), withinExpression,
```

Comment:
The actual in-memory tables (or some other DB if the user has given store annotation) is used directly to retrieve values stored in the tables. See line no 141. The inMemoryStoreCompileCondition is used to process the running aggregate data (which is not yet available in the tables). In the find method, we aggregate the running data, write it to an event chunk, and use the find method of the operator inMemoryStoreCompileCondition, to see whether it's a match.